Explore: Rich history

ivanahuckova opened the pull request:
**What this PR does / why we need it**:
This PR creates a new feature called Rich History. 

- [Design docs](https://docs.google.com/document/d/1rIi06nSoZY4d7K_3ZKM7bAEn-eBMUVu04iGHZ1M27hQ/edit)
- [Figma Design](https://www.figma.com/file/5ymYczshehFNPQIBI1fKpV/Rich-history-in-Explore?node-id=67%3A52)
- [Documentation](https://github.com/grafana/grafana/pull/22697)
- [Test coverage](https://github.com/grafana/grafana/pull/22722)
- [UX fixes and adjustments](https://github.com/grafana/grafana/pull/22729)
- [UX updates based on user feedback](https://github.com/grafana/grafana/pull/22783)

![ezgif com-video-to-gif (4)](https://user-images.githubusercontent.com/30407135/75882793-258b3300-5e22-11ea-80f3-97db89ba8e9b.gif)
![ezgif com-video-to-gif (5)](https://user-images.githubusercontent.com/30407135/75882845-4b183c80-5e22-11ea-86bb-af54f5d96f09.gif)
![ezgif com-video-to-gif (6)](https://user-images.githubusercontent.com/30407135/75883330-26709480-5e23-11ea-822d-4f0f9fd4ed6b.gif)


Rich History is under feature flag and therefore users have to add following customisation to custom.ini file: 

```[feature_toggles]
# enable features, separated by spaces
enable = richHistory
```

ivanahuckova committed:
Explore: Refactor active buttons css

ivanahuckova committed:
Explore: Add query history button

ivanahuckova committed:
WIP: Creating drawer

ivanahuckova committed:
WIP: Create custom drawer (for now)

ivanahuckova committed:
Revert changes to Drawer.tsx

ivanahuckova committed:
WIP: Layout finished

ivanahuckova committed:
Rich History: Set up boilerplate for Settings

ivanahuckova committed:
WIP: Query history cards

ivanahuckova committed:
Refactor, split components

ivanahuckova committed:
Add resizability, interactivity

ivanahuckova committed:
Save history to local storage

ivanahuckova committed:
Visualise queries from queryhistory local storage

ivanahuckova committed:
Set up query history settings

ivanahuckova committed:
Refactor

ivanahuckova committed:
Create link, un-refactored verison

ivanahuckova committed:
Copyable url

ivanahuckova committed:
Merge remote-tracking branch 'origin/master' into ivanah/rich-history

ivanahuckova committed:
Merge branch 'master' of https://github.com/grafana/grafana into ivanah/rich-history

ivanahuckova committed:
WIP: Add slider

ivanahuckova committed:
Commenting feature

ivanahuckova committed:
Slider filtration

ivanahuckova committed:
Add headings

ivanahuckova committed:
Hide Rich history behind feature toggle

ivanahuckova committed:
Cleaning up, refactors

ivanahuckova committed:
Update tests

ivanahuckova committed:
Implement getQueryDisplayText

ivanahuckova committed:
Merge remote-tracking branch 'origin/master' into ivanah/rich-history

ivanahuckova committed:
Update lockfile for new dependencies

ivanahuckova committed:
Update heading based on sorting

ivanahuckova committed:
Fix typescript strinctNullCheck errors

ivanahuckova committed:
Fix Forms, new forms

torkelo commented:
One minor (nitpick) UX thing that I think would make it cleaner is if the drawer extended fully to the edge of the page (so was flush to the edge & the menu). The content can still align with the page margins (paddings). 

thoughts @jessover9000 ? The panel editor panes go the edge of the window (as does the inspect drawer). Removes need for edge borders.

ivanahuckova commented:
> One minor (nitpick) UX thing that I think would make it cleaner is if the drawer extended fully to the edge of the page (so was flush to the edge & the menu). The content can still align with the page margins (paddings).
@torkelo that is because of the split screen view. If we would extend left drawer fully to the edge of the page, then Query History views (left and right) in the split screen would have different sizes/alignments/paddings, which probably wouldn't look that good. 

![image](https://user-images.githubusercontent.com/30407135/76075909-21842000-5f9e-11ea-9aa5-6d844b36177b.png)

aocenas commented on public/app/features/explore/RichHistory/RichHistory.tsx:
```diff
@@ -0,0 +1,219 @@
+    /* If user selects to have starredTabAsFirstTab === true, move
+     * StarredTab to first position, otherwise second
+     */
+    tabs.splice(starredTabAsFirstTab ? 0 : 1, 0, StarredTab);
```
nitpick: would probably be clearer if we just do
```
tabs = starredTabAsFirstTab ? [StarredTab, QueriesTab, SettingsTab] : [QueriesTab, StarredTab, SettingsTab]
```

ivanahuckova replied:
💙beautiful and clean solution.

aocenas commented on public/app/features/explore/RichHistory/RichHistory.tsx:
```diff
@@ -0,0 +1,219 @@
+        </TabsBar>
+        <TabContent className={styles.tabContent}>
+          {tabs
+            .filter(t => t.value === activeTab)
```
nitpick: tabs.find would probably make more sense here as we search exactly one tab.

ivanahuckova replied:
Yeah, good suggestion. Fixed! Comming in the next commit. 🙂

aocenas commented on public/app/features/explore/RichHistory/RichHistoryContainer.tsx:
```diff
@@ -0,0 +1,139 @@
+};
+
+export default hot(module)(
+  // @ts-ignore
```
Why the ts-ignore?

ivanahuckova replied:
Not needed, removed. 👍

aocenas commented on public/app/features/explore/RichHistory/RichHistoryContent.tsx:
```diff
@@ -0,0 +1,244 @@
+    retentionPeriod,
+  } = props;
+
+  const [sliderRetentionFilter, setSliderRetentionFilter] = useState<number[]>([0, retentionPeriod]);
```
I think you can use `[number, number]` as a type to specify it will always be just tuple of 2 numbers.

ivanahuckova replied:
👍

aocenas commented on public/app/features/explore/RichHistory/RichHistoryContent.tsx:
```diff
@@ -0,0 +1,244 @@
+  );
+
+  /* If in Starred tab, render following content */
+  if (onlyStarred) {
```
Feels a bit weird we have a component that basically renders 2 different things based on an if. Wouldn't it be clearer to just have 2 different components one for each tab (starred/history tab)?

ivanahuckova replied:
👍 I totally agree. The goal was to reuse as much code as possible, but it ended up as one huge component that is doing 2 separate things. I have split it into 2 components.

aocenas reviewed (commented):
Very nice overall, good work.

aocenas commented:
In addition to the comments, two general things:

1. No tests :) I think this feature is easily testable so we should add tests for this.
2. Personally it seems weird that the top tabs scroll with the content, would prefer having just the tab content scrollable.

ivanahuckova commented on public/app/core/utils/explore.ts:
```diff
@@ -336,6 +336,95 @@ export function hasNonEmptyQuery<TQuery extends DataQuery = any>(queries: TQuery
   );
 }
 
+/*
```
I agree, I have moved all richHistory-related utils to separate file  `features/explore/utils/richHistory.ts`. It is going to be included in the next commit.

ivanahuckova commented on public/app/features/explore/Explore.tsx:
```diff
@@ -348,6 +389,9 @@ export class Explore extends React.PureComponent<ExploreProps> {
                           )}
                         </>
                       )}
+                      {showRichHistory && config.featureToggles.richHistory && (
```
👍 Makes sense. I have removed this check and kept just the one next to the toggle button. Will be part of the next commit.

davkal commented:
This is going to be awesome 🚀 I only see a few blockers before this could be shipped from a feature point of view. Here some high level user feedback:

Blocker:
- I could not load the query from a history, other than copying it and putting it into the query field again. Clicking on a history entry did not do anything.
- the action icons in the item rows need some tooltip or hinting
- button to clear the history with one click for the paranoid

Minor:
- the drawer left and right limits coincide with the graph/table/cheatsheet width. This looks weird because it breaks any sort of physical drawer metaphor where a drawer either overlaps with what is underneath (e.g., top drawer of a chest of drawers) or is a sort of help area that's smaller than the main area (e.g., a pull-out writing surface of a writer's desk).
- the list of datasources should only show entries that have items in the history, ideally it shows the number of entries too, so I know which DS i was using most
- the history slider needs step adapted to the complete range. E.g., 7 steps are fine for a week, but 2 steps for last 2 days are weird and are missing an opportunity for higher resolution filtering. For 2 days I would expect 6h steps.
- as long as the history is per browser, we should mention this. Perhaps below the list in fine print: "Query history is a beta feature. The history is local to your browse and is not shared with others. Feedback?" with a link to a github issue.

ivanahuckova committed:
Fixes based on provided feedback

oddlittlebird commented:
Where are the docs?

ivanahuckova commented:
@aocenas 
> 1. No tests :) I think this feature is easily testable so we should add tests for this.

I know 🙈 🙈 🙈. I know that it is not the best practise, but could I maybe do tests for this feature in another PR, as this PR on its own is pretty big. 

> 2. Personally it seems weird that the top tabs scroll with the content, would prefer having just the tab content scrollable.

Yes you are right, fixed!

ivanahuckova commented on public/app/features/explore/RichHistory/RichHistory.tsx:
```diff
@@ -0,0 +1,219 @@
+  Settings = 'Settings',
+}
+
+export const RICH_HISTORY_SETTING_KEYS = {
```
👍

jessover9000 commented:
It's already looking awesome @ivanahuckova 🔥 🚀 ✨ now for the small details! :)

> > One minor (nitpick) UX thing that I think would make it cleaner is if the drawer extended fully to the edge of the page (so was flush to the edge & the menu). The content can still align with the page margins (paddings).
> 
> @torkelo that is because of the split screen view. If we would extend left drawer fully to the edge of the page, then Query History views (left and right) in the split screen would have different sizes/alignments/paddings, which probably wouldn't look that good.

I designed it without borders in Figma. https://www.figma.com/file/5ymYczshehFNPQIBI1fKpV/Rich-history-in-Explore?node-id=101%3A318 I'm not sure what you mean about "different paddings", do you just mean the different alignments if you resize the views differently with the handlebars? Because I think that's a natural part of the interface and I don't mind it. I would make it full width. But maybe I'm not understanding correctly?

Good suggestion on the Clear History button @davkal , +1 on that.

I have quite a bunch of additional feedback, not sorted by priority but by area. some are more nitpicky than others, but I'm sure you can figure that out by yourself 😉

**general display/outside of history panel**
- [ ] screen width: space between graph view and history / scrolling doesn’t work even though panel is cut off
- [ ] border-bottom for history panel should be gone
- [ ] add a little animation for when you toggle that settings button and query history/starred tabs switch positions / soften the switch a little
- [ ] Handlebar has no color contrast to cards/panels and fades into them depending on view options (see screenshot) – maybe give the handlebar a drop shadow
- [ ] there is a separate github issue for the active toggle button border color: I think it would be nice to change it to the lighter yellow https://github.com/grafana/grafana/issues/21958
- [ ] I think we should introduce a minimum height for the history, maybe around 200px? or below that height the panel should be more responsive (smaller font sizes, tab/input heights, etc.) so that it stays usable. currently if it gets too narrow I feel it's hard to interact with it.

**query history list and cards**
- [ ] headlines/session titles: would love a smaller number of queries indicator “X queries” next to the session headlines in body text styling (cf. figma: https://www.figma.com/file/5ymYczshehFNPQIBI1fKpV/Rich-history-in-Explore?node-id=99%3A538)
- [ ] Edit icon next to session titles to inline edit the titles
- [ ] currently the history cards/query texts on the cards are not clickable. would make them clickable and on click the Explore query should be replaced by the history item but without running it
- [ ] if I have a history item with several items, the border separating them is supposed to be only 1px
- [ ] history card text alignment: currently the paddings are equal on all sides but opticallly, it looks as if there’s more space in the bottom than in the top. would recommend decreasing the bottom padding by 2px as long as there’s no comment
- [ ] drop shadow in dark theme is too light: should be blur: 4px, y: 2px, pure black opacity 100%

**filtering**
- [ ] please align the slider headline “Filter history between” to the top of the Filter and Sort dropdowns
- [ ] if “today” and “a week ago” can’t be positioned to the side of the slider, there should be more spacing between “Filter history between” and the slider label – increase margin-bottom to 12px
- [ ] the time filter slider doesn't resize with the window when you use the handlebar – I think it would be nice if it would do that as much as possible
- [ ] when I add a filter and remove it again I can’t see queries any more
- [ ] When I hide and show the query history, the filters are reset – maybe we can save the last settings the user had? (if you think that makes sense)

**commenting**
- [ ] Add comment form input should be styled (currently it has no border and no rounded corners). maybe add resizing handle to the input in case people want to write longer text
- [ ] maybe move save/cancel button to the right of the comment input on history panel widths >=768px to save vertical space
- [ ] comment text looks identical to query text, is really hard to tell apart. if you check out figma, the query text is supposed to be body text (14px) medium and the comment is 12px regular.
- [ ] add margin-top: 4px between comment text and queries

ivanahuckova committed:
Fixes, splitting component into two

ivanahuckova committed:
Add tooltips, add delete history button

ivanahuckova committed:
Clicking on card adds queries to query rows

ivanahuckova committed:
Delete history, width of drawers

ivanahuckova commented:
@davkal 
> Blocker:
> 
> * I could not load the query from a history, other than copying it and putting it into the query field again. Clicking on a history entry did not do anything.

[Fixed here](https://github.com/grafana/grafana/pull/22570/commits/f3c05cbf78fa55ae0ae39893ca6a1f70aa19e30e#diff-02381711eae3a10406c1c2a6fa529f99R101-R120)

> * the action icons in the item rows need some tooltip or hinting

[Fixed here](https://github.com/grafana/grafana/pull/22570/commits/006369e6f033b8963be0315dc9d9ed50ed377036#diff-02381711eae3a10406c1c2a6fa529f99R142-R169)

> * button to clear the history with one click for the paranoid

[Fixed here](https://github.com/grafana/grafana/pull/22570/commits/44fd8691ca3ce3ec7caaa110cf9d9fd9c448fd57#diff-a1d51f428ea85adf8a04ce4dfa2ce217R101-R107)
![image](https://user-images.githubusercontent.com/30407135/76145377-4acdaa80-6089-11ea-8367-992018d0f902.png)

> Minor:
> 
> * the drawer left and right limits coincide with the graph/table/cheatsheet width. This looks weird because it breaks any sort of physical drawer metaphor where a drawer either overlaps with what is underneath (e.g., top drawer of a chest of drawers) or is a sort of help area that's smaller than the main area (e.g., a pull-out writing surface of a writer's desk).

Fixed
![image](https://user-images.githubusercontent.com/30407135/76145402-751f6800-6089-11ea-9a13-a5fb4e3a372b.png)

> * the list of datasources should only show entries that have items in the history, ideally it shows the number of entries too, so I know which DS i was using most

Fixed 
![image](https://user-images.githubusercontent.com/30407135/76145420-96805400-6089-11ea-8f5e-bf8052033e62.png)

> * the history slider needs step adapted to the complete range. E.g., 7 steps are fine for a week, but 2 steps for last 2 days are weird and are missing an opportunity for higher resolution filtering. For 2 days I would expect 6h steps.

Not fixed yet

> * as long as the history is per browser, we should mention this. Perhaps below the list in fine print: "Query history is a beta feature. The history is local to your browse and is not shared with others. Feedback?" with a link to a github issue

Fixed 
![image](https://user-images.githubusercontent.com/30407135/76145420-96805400-6089-11ea-8f5e-bf8052033e62.png)

ivanahuckova commented:
**✅FIXED:**

> [x] - One minor (nitpick) UX thing that I think would make it cleaner is if the drawer extended fully to the edge of the page (so was flush to the edge & the menu). The content can still align with the page margins (paddings).
 
@torkelo @jessover9000 Sorry I totally miss-understood/miss-read. I thought we are talking about drawer that goes over the navigation on the left side. But now I understand what do you mean. Yes I agree, this has been fixed. 
![image](https://user-images.githubusercontent.com/30407135/76205723-1d087300-61fb-11ea-9b9f-c635434fca1b.png)

>  [x] - Good suggestion on the Clear History button @davkal , +1 on that.

Fixed (see comment to @davkal feedback)

> * [x]  there is a weird unfinished corner in the right border of the history window after using the scroll bar, not sure if that can be fixed but it looks strange

Fixed (see the screenshot for fully extended drawer)

> * [x]  border-bottom for history panel should be gone

Removed, there is currently only border on top and I have removed other borders, as they are not needed anymore, as drawer is extended fully to the corners.  

> * [x]  padding-top of the history area under the tabs can be 12px instead of 24px, don't need to waste space here ;)

Fixed to 16px (@jessover9000 let me know if you would like to change it to 8px instead, but in our spacing variables that we use, we have either 8px or 16px spacing):
![image](https://user-images.githubusercontent.com/30407135/76209886-3c0b0300-6203-11ea-90e0-506b81b1fee9.png)

> * [x]  **currently the history cards/query texts on the cards are not clickable. would make them clickable and on click the Explore query should be replaced by the history item but without running it** (this addresses the blocker David mentioned)

Fixed (see comment to @davkal feedback)

> * [x]  if I have a history item with several items, the border separating them is supposed to be only 1px

Fixed - changed to 1px

> * [x]  drop shadow in dark theme is too light: should be blur: 4px, y: 2px, pure black opacity 100%

Fixed, changed to `0px 2px 4px black` for dark theme.

> * [x]  please align the slider headline “Filter history between” to the top of the Filter and Sort dropdowns

Fixed

> * [x] “Filter history between” and the slider label – increase margin-bottom to 12px

Changed to 16px - same reason as above. Let me know if you would like me to change it to 8px

> * [x]  comment text looks identical to query text, is really hard to tell apart. if you check out figma, the query text is supposed to be body text (14px) medium and the comment is 12px regular.

Changed to 12px

> * [x]  add margin-top: 4px between comment text and queries

Fixed

**❓NEED MORE INFO:**

> * [ ]  screen width: space between graph view and history / scrolling doesn’t work even though panel is cut off (see screenshot) – I'd expect the main area above the panel to be scrollable still

Content behind the drawer is scrollable, if there is a content that can be scrolled. I am not sure if you are suggesting, that we should resize the explore screen when we open the drawer (like in panel edit - pulling up the edit panel makes visualisation panel smaller)? 

> * [ ]  Handlebar has no color contrast to cards/panels and fades into them depending on view options (see screenshot) – maybe give the handlebar a drop shadow

I have checked other handle bars and they don't have drop shadow. I have also played around with it and coudn't find anything that I like. Do you have any specific suggestion for box shadow?  

> * [ ]  I think we should introduce a minimum height for the history, maybe around 200px? or below that height the panel should be more responsive (smaller font sizes, tab/input heights, etc.) so that it stays usable. currently if it gets too narrow I feel it's hard to interact with it.

Can we user test this and see if it is the issue? My personal opinion:
- Setting minimum height is not necessary. When history drawer is opened, the size is appropriate for interaction with history. If user wants to temporary make it very small//hide it/is currently not interacting with it, because is interacting with Explore, we should let him. If they want to interact with it, they will pull it up to appropriate size. 
- Responsive sizing - Again, I don't see any reason for resizing. I also think, it would be hard to make it look good (not just suddenly skip to small size), with almost no additional value for user. If we are concerned about sizing, we could make everything defaultly smaller, but again, I personally don't see any issue with this and I would rather see if users give us a feedback on this.

However, if have a strong opinion about this, let me know and I will introduce minimal height.

> * [ ]  history card text alignment: currently the paddings are equal on all sides but opticallly, it looks as if there’s more space in the bottom than in the top. would recommend decreasing the bottom padding by 2px as long as there’s no comment. 

Would it be okay to do decrease by 4px (So from 8px to 4px - see image below). Because if yes, we can use predefined padding `spacing.xs` instead of `spacing.sm`. Otherwise, we would have to customize it and implicitly specify that padding (to 6px), which I would prefer not to do. Or we could set it to 4px defaultly (`spacing.xs`) and without comment, use 2px padding(`spacing.xxs`).

So what I am trying to say - we have these spacing variables and I would prefer to use them, than just harcoding values:
![image](https://user-images.githubusercontent.com/30407135/76214418-fb17ec00-620c-11ea-9141-069f69241636.png)

![image](https://user-images.githubusercontent.com/30407135/76210824-31ea0400-6205-11ea-861e-5bf6c8bc250e.png)

> * [ ]  when I add a filter and remove it again I can’t see queries any more

Can you please try this again. I have tested several times and it works for me - removing filters results in showing all queries. Or maybe make a video/gif - so I can simulate.

**🆕SHOULD BE PART OF NEW/SEPARATE PR - NOT EXCLUSIVELY CONNECTED TO THIS PR: **

> * [ ]  there is a separate github issue for the active toggle button border color: I think it would be nice to change it to the lighter yellow #21958

**🤸‍♂️NEXT ITERATION: **

> * [ ]  Edit icon next to session titles to inline edit the titles

I have not implemented sessions. This needs more work.


**👩‍💻WIP (WORK IN PROGRESS, CURRENTLY WORKING ON: **
> * [ ]  headlines/session titles: would love a smaller number of queries indicator “X queries” next to the session headlines in body text styling

> * [ ]  the time filter slider doesn't resize with the window when you use the handlebar – I think it would be nice if it would do that as much as possible

> * [ ]  When I hide and show the query history, the filters are reset – maybe we can save the last settings the user had? (if you think that makes sense)

> * [ ]  Add comment form input should be styled (currently it has no border and no rounded corners). maybe add resizing handle to the input in case people want to write longer text

> * [ ]  maybe move save/cancel button to the right of the comment input on history panel widths >=768px to save vertical space

> * [ ]  add a little animation for when you toggle that settings button and query history/starred tabs switch positions / soften the switch a little

> * [ ]  if “today” and “a week ago” can’t be positioned to the side of the slider, there should be more spacing between

ivanahuckova committed:
UI/UX changes, updates

jessover9000 commented:
> Content behind the drawer is scrollable, if there is a content that can be scrolled. I am not sure if you are suggesting, that we should resize the explore screen when we open the drawer (like in panel edit - pulling up the edit panel makes visualisation panel smaller)?

Ok, then it isn't working for me 🤔I didn't test it with very big content, but I've definitely had cases where the content behind the drawer was a little too big and a piece of it vanished and then I couldn't scroll. I didn't get a scrollbar and moving my cursor above the rich history and using my mousewheel didn't trigger any scrolling for me either. 

> I have checked other handle bars and they don't have drop shadow. I have also played around with it and coudn't find anything that I like. Do you have any specific suggestion for box shadow?

In the design library in Figma, I've only used two types so far, one just like the one you implemented for cards and one less soft version. So maybe just try the one for cards and see whether it looks ok or whether it's too much blur?

> Would it be okay to do decrease by 4px (So from 8px to 4px - see image below).

yes that looks fine :)

> Can we user test this and see if it is the issue?

I did feel like when you make the history very small it can look like an error/like something is wrong. But I guess you are right, we can leave it for now and see whether people mind it. There will be user tests at GrafanaCon after all ;)

I will test the filtering again. Good work, @ivanahuckova 💪 ✨ 💃

ivanahuckova committed:
Add number of queries to headings, box shadows

oddlittlebird commented:
Per my discussion with @ivanahuckova earlier today:
* Tab and button text needs to be sentence case.
* "Data sources" needs to be two words in the UI, not one.

ivanahuckova committed:
Fix slider, remove feature toggle

ivanahuckova commented:
@jessover9000 @davkal @aocenas 

These are the unresolved issues:

- [ ] When I hide and show the query history, the filters are reset – maybe we can save the last settings the user had? (if you think that makes sense)
- [ ] the time filter slider doesn't resize with the window when you use the handlebar – I think it would be nice if it would do that as much as possible 
- [ ] maybe move save/cancel button to the right of the comment input on history panel widths >=768px to save vertical space
- [ ] add a little animation for when you toggle that settings button and query history/starred tabs switch positions / soften the switch a little
- [ ] if “today” and “a week ago” can’t be positioned to the side of the slider, there should be more spacing between
- [ ] Add test coverage
- [ ] Add error handling

As the branch of is on wednesday (and tomorrow is the last possible moment we could merge this branch), let me know if any of these issues are blocking merge/release (keep in mind we will present this as a beta feature). If not, could you please review and approve/disapprove.

oddlittlebird commented:
Ivana, I would recommend not adding screenshots to documentation until the
location of buttons and such is absolutely, totally final. This will save
you rework later, and I would argue that lots of screenshots aren't
necessary for beta documentation.

On Mon, Mar 9, 2020 at 12:19 PM Ivana Huckova <notifications@github.com>
wrote:

> @jessover9000 <https://github.com/jessover9000> @davkal
> <https://github.com/davkal> @aocenas <https://github.com/aocenas>
>
> These are the unresolved issues:
>
>    - When I hide and show the query history, the filters are reset –
>    maybe we can save the last settings the user had? (if you think that makes
>    sense)
>    - the time filter slider doesn't resize with the window when you use
>    the handlebar – I think it would be nice if it would do that as much as
>    possible
>    - maybe move save/cancel button to the right of the comment input on
>    history panel widths >=768px to save vertical space
>    - add a little animation for when you toggle that settings button and
>    query history/starred tabs switch positions / soften the switch a little
>    - if “today” and “a week ago” can’t be positioned to the side of the
>    slider, there should be more spacing between
>    - Add test coverage
>    - Add error handling
>
> As the branch of is on wednesday (and tomorrow is the last possible moment
> we could merge this branch), let me know if any of these issues are
> blocking merge/release (keep in mind we will present this as a beta
> feature). If not, could you please review and approve/disapprove.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/grafana/grafana/pull/22570?email_source=notifications&email_token=AMNF6KOBT7XJYWLDZTIOPXDRGU6MXA5CNFSM4LBHBAEKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEOIWUII#issuecomment-596732449>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMNF6KN5VLMCZTHLKX3U4ALRGU6MXANCNFSM4LBHBAEA>
> .
>


-- 
Diana Payton (She/Her)
Sr. Technical Writer at Grafana Labs

ivanahuckova committed:
Fix typo in the beta announcement

ivanahuckova committed:
Fix how the rich history state is rendered when initialization

ivanahuckova committed:
Move updateFilters when activeDatasourceFilter onlyto RichHistory, remove duplicated code

ivanahuckova commented on public/app/features/explore/RichHistory/RichHistoryStarredTab.tsx:
```diff
@@ -0,0 +1,150 @@
+   *  Filtering based on datasource won't be available. Otherwise set to null, as filtering will be
+   * available for user.
+   */
+  useEffect(() => {
```
fixed in https://github.com/grafana/grafana/pull/22570/commits/65f20cb07c968f582726bc91ad4ef9715c63db9f

aocenas reviewed (approved):
Looks good now. There are some issues with the build though.

ivanahuckova committed:
Fix typescript strictnull errors, not used variables errors

torkelo commented:
Sorry to be a little critical, but a big feature like this without any tests feels bad. This should have at least some unit tests and a end-2-end smoke test would have been amazing. If that means it gets pushed to the next release that is fine. I do not think waiting to write unit tests in a follow-up PR is a good strategy. 

In the future implement big new features like this behind a feature toggle then it's easier to do things in smaller PRs and get UX feedback.

ivanahuckova commented:
@torkelo No, you are right. I totally agree. I talked to @aocenas and he was saying basically the same thing. I thought that I would have more time to implement unit tests before 6.7-beta, but unfortunately I didn't. So I thought that feature is going to be ready and fully tested for 6.7 release.

I am also totally aware that I should have split this into more smaller PRs (have it under feature flag), write unit tests during the coding. It was a big feature and I didn't realize at the beginning that it is going to be that big and complex (even though we have really thoroughly planned it). That was my lack of experience with bigger feature and lack of asking for help/coordination. 🙈

On the other side, we wanted to have this first iteration out in 6.7 as a beta feature, and let user know that it is a beta feature and ask for their feedback (it is also mentioned directly in the Rich History). So we would have that feedback as soon as possible and iterate based on their feedback. 

So what I am trying to say - I have removed the 6.7-beta milestone and striked trough the Rich History information from What's new in 6.7. If you think we should postpone it and have it in 7 when it is ready and fully tested, I totally agree. It will give us/me time to add unit tests and work on feedback from Jess, Andrej and David.

torkelo commented:
Understand @ivanahuckova great that you have those discussions, no worries was just making sure we learn from this so it does not become the norm :) 

> On the other side, we wanted to have this first iteration out in 6.7 as a beta feature, and let user know that it is a beta feature and ask for their feedback (it is also mentioned directly in the Rich History). So we would have that feedback as soon as possible and iterate based on their feedback. 

Sounds good, early feedback is great. Think we can mention it as a beta feature in the what's new docs if we want to highlight it and ask for feedback.

ivanahuckova commented:
Absolutely! 🙂I'll make sure that this feature will be fully tested before the end of tomorrow and I will make sure that all of my features have tests as a part of PR from now on. 

And just to make sure - are you okay with Rich History being part of 6.7-beta release as a beta feature?
